internal/uvm/virtual_device.go

```diff
@@ -0,0 +1,39 @@
+package uvm
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/Microsoft/hcsshim/internal/requesttype"
+	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
+)
+
+const (
+	resourcePathVpciFmt = "VirtualMachine/Devices/VirtualPci/%s"
+)
+
+func (uvm *UtilityVM) GpuCounter() uint64 {
+	return atomic.AddUint64(&uvm.gpuCounter, 1)
+}
+
+func (uvm *UtilityVM) AssignDevice(ctx context.Context, id string, device hcsschema.VirtualPciDevice) error {
+	uvm.m.Lock()
+	defer uvm.m.Unlock()
+
+	return uvm.modify(ctx, &hcsschema.ModifySettingRequest{
+		ResourcePath: fmt.Sprintf(resourcePathVpciFmt, id),
+		RequestType:  requesttype.Add,
+		Settings:     device,
```

Comment:
Is there no guest side operation to do here?

Reply:
not for gpus. The only thing the guest really handles is setting up the prestart hook on the container spec before going to runc (note: prestart hook has been deprecated and I'll update the code when we have the new release of runc in the uvm) This might need to change for other vpci devices but it's unclear since I haven't played around with those yet, so for now just took out the guest side.

Reply:
I imagine that the device just shows up, available for use with the CLI tool... but we are probably going to hit race conditions if we don't have some kind of guest request here to wait for the device in some way.

Reply:
Seems like if we verify the UUID match then we can just wait for it to show up at least.

Reply:
Updated to add a guest side request to wait for the device's vpci bus location directory under /sys/bus/vmbus/devices/<vmbusUuid>/...